### ISSUE TITLE
Yubikey 4 PIV certificates

### ISSUE THREAD

**Issue by wiktor-k** [NoMoreFood/putty-cac#7]
Hello,

Did anyone try Yubikey 4 as a smartcard holding the certificates? Because for me currently neither Putty CAC or Pageant CAC show Yubikey-stored certificates in certificate selection dialog. Only the software certificates are shown.

Yubikey 4 certs are displayed in `certmgr.msc` and also work correctly in Chrome/Edge (HTTPS client authentication) and in curl built with WinSSL. Putty is the only "missing link" for me.

I tried some changes from #2 (like changing chuid etc.) but no dice.

**Comment by NoMoreFood** [comment]
Are these present in the user certificate store or the system certificate store?  Do you have an export of a .cer file that you can email me?

**Comment by wiktor-k** [comment]
User certificate store under Personal, they are placed there automatically by Windows when the card is first time detected.

I've e-mailed you the cer file.

The cert is signed by my custom test CA but having CA trusted or not in cert store did not influence putty.

Thanks for help!

**Comment by wiktor-k** [comment]
Now that I'm looking at the diff of certificates in store and what is presented by Putty I see also a S/MIME e-mail certificate (issued by COMODO) that Putty does not see. But looking [at the code here](https://github.com/NoMoreFood/putty-cac/blob/1b372260639a2fc8ac14758edd6a850b8ccdfa29/cert/cert_capi.c#L160-L161) does not seem like Putty is putting any kind of filter when requesting the certs list.

**Comment by NoMoreFood** [comment]
@wiktor-k I believe I have located the problem.  More to come...

**Comment by NoMoreFood** [comment]
@wiktor-k Did you get my email with the new binaries to test?

**Comment by wiktor-k** [comment]
Yes, or no. I mean I got it but it was in a spam folder and I haven't seen it (too agressive spam filter I guess, even with the rename trick).

But I just tried putty from your bundle and *IT WORKS!*. Cert is visible and cert is usable. I just logged in using the smartcard to an OpenSSH server. Wow... great!

I haven't tried Pageant CAC yet but I will next week, I suppose it will work though...

Thanks for super quick help and solving this issue, CAPI certs for SSH was  an only missing block for me.

**Comment by NoMoreFood** [comment]
@wiktor-k  Great! Let me know how Pageant works and if all looks good, I'll make an "official" release next week.  The version I sent you is compiled the same way as any official release so no worries about continuing to use it.

**Comment by wiktor-k** [comment]
Okay, I've verified Pageant CAC (both x86 and x64 versions) and it also works perfectly. With multiple versions of Putty (even the official one).

I've also checked `ssh-pageant` (proxy for OpenSSH to Pageant) and it also works (although the PIN dialog in this case doesn't get the focus but is opened in the background :-/ ) With Putty focus works correctly but I guess the `ssh-pageant`-related issue is outside of the scope of this project so it's just an observation.

Another thing to note is that smartcard PIN is requested on each connection even with "once per session" PIN policy set. In the browser the PIN is cached for a very long time. I don't know if it's something that could be adjusted but I thought Pageant would cache it (similarly to how GPG4Win asks for PIN only when the card is inserted). I don't know if that's something that is possible for Pageant to do but I'd be *very* glad if PIN was cached (it's not a security issue because I'll be using touch-to-use also).

One way or another the cert is usable and everything works. Thanks a lot! 👍

**Comment by NoMoreFood** [comment]
@wiktor-k Thank you for the feedback.  Did you try the 'Force PIN Caching' option in the Pageant system tray menu?

**Comment by wiktor-k** [comment]
@NoMoreFood eeerrm... nope I didn't. I didn't even know that option was there :)

Now I used it and even though the PIN dialog is different **it works correctly**. Verified with multiple Putties, `ssh-pageant` etc. I'm very glad!

Note that the PIN is cached even when the smartcard is ejected. That's different than GPG4Win's caching that is cleared when card is removed but it's not a big problem for me (I'm just noting that for future reference).
